### ISSUE TITLE
Invalid Go code is generated in pulumi-aws

### ISSUE THREAD

**Issue by stack72** [pulumi/pulumi#7623]
This is not a new issue but we have not been building the generated go sdk for the pulumi-aws provider. After removing all of the go sdk and letting it be regenerated, we are actually generating invalid code:

```
# github.com/pulumi/pulumi-aws/sdk/v4/go/aws/kinesisanalyticsv2
go/aws/kinesisanalyticsv2/pulumiTypes.go:2388:43: impossible type assertion:
	ApplicationApplicationConfigurationSqlApplicationConfigurationOutput does not implement pulumi.Output (missing ApplyT method)
go/aws/kinesisanalyticsv2/pulumiTypes.go:2396:43: impossible type assertion:
	ApplicationApplicationConfigurationSqlApplicationConfigurationOutput does not implement pulumi.Output (missing ApplyT method)
go/aws/kinesisanalyticsv2/pulumiTypes.go:2459:11: cannot use v.Input (type *ApplicationApplicationConfigurationSqlApplicationConfigurationInput) as type *ApplicationApplicationConfigurationSqlApplicationConfigurationInput in return argument:
	*ApplicationApplicationConfigurationSqlApplicationConfigurationInput is pointer to interface, not interface
go/aws/kinesisanalyticsv2/pulumiTypes.go:2466:11: cannot use v.Outputs (type []ApplicationApplicationConfigurationSqlApplicationConfigurationOutput) as type []ApplicationApplicationConfigurationSqlApplicationConfigurationOutput in return argument
go/aws/kinesisanalyticsv2/pulumiTypes.go:2503:11: cannot use v.Input (type *ApplicationApplicationConfigurationSqlApplicationConfigurationInput) as type *ApplicationApplicationConfigurationSqlApplicationConfigurationInput in return argument:
	*ApplicationApplicationConfigurationSqlApplicationConfigurationInput is pointer to interface, not interface
go/aws/kinesisanalyticsv2/pulumiTypes.go:2513:11: cannot use v.Outputs (type []ApplicationApplicationConfigurationSqlApplicationConfigurationOutput) as type []ApplicationApplicationConfigurationSqlApplicationConfigurationOutput in return argument
go/aws/kinesisanalyticsv2/pulumiTypes.go:2527:6: ApplicationApplicationConfigurationSqlApplicationConfigurationInput redeclared in this block
	previous declaration at go/aws/kinesisanalyticsv2/pulumiTypes.go:2363:6
go/aws/kinesisanalyticsv2/pulumiTypes.go:4498:6: ApplicationApplicationConfigurationSqlApplicationConfigurationOutput redeclared in this block
	previous declaration at go/aws/kinesisanalyticsv2/pulumiTypes.go:2432:6
go/aws/kinesisanalyticsv2/pulumiTypes.go:6712:96: cannot use ApplicationApplicationConfigurationSqlApplicationConfigurationOutput{} (type ApplicationApplicationConfigurationSqlApplicationConfigurationOutput) as type pulumi.Output in argument to pulumi.RegisterOutputType:
	ApplicationApplicationConfigurationSqlApplicationConfigurationOutput does not implement pulumi.Output (missing ApplyT method)
```

I am aiming to include our go sdk build as part of our CI process but this will fail until we actually have this codegen fixed

You can see this failure locally using the command:

```
cd sdk/ && GOGC=50 go list github.com/pulumi/pulumi-aws/sdk/v4/go/aws/... | xargs go build
```

**Comment by emiliza** [comment]
Should be fixed by https://github.com/pulumi/pulumi/issues/7595

**Comment by EvanBoyle** [comment]
Unfortunately this error is not fixed by #7595. This is a pre-existing issue dating back 10 months looking at history of the generated code. 

The issue suffix collision, the same as the one described here https://github.com/pulumi/pulumi/issues/7622#issuecomment-901448822

https://github.com/pulumi/pulumi-aws/blob/05f798acb396494512204c48fe57bf2c2d7086f8/sdk/go/aws/kinesisanalyticsv2/pulumiTypes.go#L2585
https://github.com/pulumi/pulumi-aws/blob/05f798acb396494512204c48fe57bf2c2d7086f8/sdk/go/aws/kinesisanalyticsv2/pulumiTypes.go#L2417

There are two types in the schema: `T` and `TInput` and for the first type we generate `type TInput interface` which collides with the second types `type TInput struct`.
